### ISSUE TITLE
@0 annotation on lambda ignored

### ISSUE THREAD

**Issue by nad** [agda/agda#5251]
The following code is accepted:
```agda
_ : @ω Set → Set₁
_ = λ (@0 _) → Set
```
Given that we no longer (by default) have subtyping for `@0` (#4474) I would assume that this code should be rejected.

**Comment by nad** [comment]
This should be easy to fix:

https://github.com/agda/agda/blob/4773470ceb0e8671e128109f4728d6416ff93fae/src/full/Agda/TypeChecking/Rules/Term.hs#L606-L619

**Comment by nad** [comment]
The fix referred to above replaces `moreQuantity` with `sameQuantity`. However, I guess that if `--subtyping` is on, then the code should be allowed. Then we get another problem, the code is rejected if `--double-check` is used:

https://github.com/agda/agda/blob/465e46a20b59160ea38bd50a30beaab849d176c5/src/full/Agda/TypeChecking/CheckInternal.hs#L310-L321

@jespercockx, should the double-checker use `moreUsableModality` if `--subtyping` has been turned on?

**Comment by jespercockx** [comment]
Yes, that seems plausible.

**Comment by nad** [comment]
This procedure is used both for lambdas and for applications, so I guess one should use either `moreUsableModality` or `flip moreUsableModality`.

An alternative might be to turn off subtyping completely for erasure and relevance. Does anyone rely on it?

**Comment by nad** [comment]
I tried to fix the issue with `--subtyping`, and the output for `Issue3855d` changed:
* Before:
  ```
  Issue3855d.agda:26,1-30,4
  Expected an erased argument, but found a unrestricted argument
  when checking that the clause
  works = record { to = [_] ; from = _ ; to∘from = λ _ → refl } has
  type {A : Set} → A ↠ Erased A
  ```
* After:
  ```
  Failed to solve the following constraints:
    _from_12 x = erased x : A (blocked on _from_12)
  Unsolved metas at the following locations:
    Issue3855d.agda:28,15-16
    Issue3855d.agda:29,21-25
  ```

This reverses the change that accompanied @jespercockx's fix of issue #4727 (fa0dfc466f7ecf7a4e8cdbc3afacb86afcc85a8a). I wonder if this indicates that my "fix" broke something.

Given that `--subtyping` is off by default (and hence presumably not used a lot), and that, in @jespercockx's [words](https://github.com/agda/agda/issues/4474#issue-569515513), "it seems that we do not yet fully understand how constraint solving and metavariables […] are supposed to work" (in a setting with subtyping), I suggest that we disable subtyping for erasure and relevance. Does anyone disagree with this?

**Comment by jespercockx** [comment]
It's already hidden under a flag, I do not see the need to disable it completely. We have plenty of other options that we do not understand fully but are still there hidden under a flag (--experimental-irrelevance comes to mind, --sized-types might qualify as well).

I'd rather just move it to "currently unsupported" status than remove it completely.

**Comment by nad** [comment]
> I'd rather just move it to "currently unsupported" status than remove it completely.

In that case I think we should also make it incompatible with `--safe`.

**Comment by jespercockx** [comment]
I wouldn't mind that.

**Comment by nad** [comment]
Note that `--cumulativity` implies `--subtyping`. Are you fine with making `--cumulativity` incompatible with `--safe`?

**Comment by jespercockx** [comment]
The original reason for introducing --subtyping was to avoid a performance problem with a certain check in the constraint solver that's only required when subtyping is used. But at the moment it is also the only way to turn on or off the specific subtyping rules for irrelevance and erasure. This has the weird effect that turning on --cumulativity also automatically turns on those other subtyping rules. Perhaps it would be better to introduce a new more specific flag (`--modality-subtyping` perhaps?) for turning on those rules, and make just that one incompatible with `--safe` (but not `--subtyping` itself).

**Comment by nad** [comment]
I don't think we should spend time on a feature that no one is using. What about removing `--subtyping` entirely, and only keeping those parts of the implementation that are related to `--cumulativity`?

**Comment by jespercockx** [comment]
Well, I'm not 100% sure on the "no one is using" it part. But we could do the experiment by removing it just after the release, and seeing if anyone complains.

**Comment by nad** [comment]
I've fixed the problem in the OP. The fix might break some code that uses `--subtyping` and `λ (@0 x) → …`, but I doubt that there is a lot of code like that in use.

I plan to merge the fix into the `release-2.6.2` branch.

**Comment by nad** [comment]
I opened a new issue for the removal of support for subtyping for erasure and irrelevance.
